bin/studio.py

```diff
@@ -0,0 +1,276 @@
+        self.path = path
+        self.method = method
+        self.data = data
+        self.public_api_client = public_api_client
+
+        self.path_params = [
+            parameter[1:-1]
+            for parameter in self.path.split("/")
+            if (parameter and parameter[0] == "{")
+        ]
+        path_entities = [
+            parameter
+            for parameter in path.split("/")
+            if (parameter and parameter[0] != "{")
+        ]
+
+        self.name = self.create_command_name(method, path_entities)
+
+    def is_singular(self):
+        if self.method == "post":
+            # /media/permissions or /collections/permissions
+            if self.path.endswith("permissions"):
+                return False
+
+            return True
+
+        if self.method != "get":
+            return False
+
+        # "/media/{media_id}/download": {
+        if "200" not in self.data["responses"]:
+            return False
+
+        return_ok = self.data["responses"]["200"]
+
+        # /perspectives/{perspective_id}/insights/overview
+        if "schema" not in return_ok:
+            return False
+
+        #     "/media/search": {
+        schema = return_ok["schema"]
+        props = schema.get("properties", None) or schema["allOf"][0].get(
+            "properties", None
+        )
+
+        # "/collections/{collection_id}": {
+        return_types = [v.get("type", "") for k, v in props.items()]
+
+        return any([x != "array" for x in return_types])
+
+    def create_command_name(self, method, path_entities):
+        if path_entities[-1] in ["search", "download", "complete", "ping", 'transfer_media']:
+            return "_".join([path_entities[-1]] + path_entities[:-1])
+        else:
+            command_name = (
+                f"{Command.method_to_command[method]}_{'_'.join(path_entities)}"
+            )
+            if command_name[-1] == "s" and self.is_singular():
+                return command_name[:-1]
+            else:
+                return command_name
+
+    def parameter_names(self):
+        return [p["name"] for p in self.data.get("parameters", {})]
+
+    def is_csv_command(self):
+        return self.name in [
+            "show_perspectives_insights_overview",
+            "show_perspectives_insights_users",
+        ]
+
+    def execute(self, args):
+        params = {
+            k: v
+            for k, v in vars(args).items()
+            if (k in self.parameter_names() and k not in self.path_params)
+        }
+
+        response = self.public_api_client.request(
+            self.method, self.path.format(**vars(args)), params=params
+        )
+
+        if response.ok:
+            # can be 'application/json; charset=utf-8' or 'video/mp4'
+            content_type = response.headers["Content-Type"].split(";")
+            if content_type[0] == "application/json":
+                return json.dumps(response.json(), indent=2)
+            elif self.name.startswith("download_"):
```

Comment:
Since every elif ends with a return, I would probably change it to "early returns" with blank lines between:

```
if CASE_1:
    ...
    return SOMETHING_1

if CASE_2:
    ...
    return SOMETHING_2

...

return FINAL_SOMETHING
```

Reply:
good catch, fixing it.